forc/src/utils/helpers.rs

```diff
@@ -52,10 +53,73 @@ pub fn read_manifest(manifest_dir: &Path) -> Result<Manifest, String> {
             ))
         }
     };
-    match toml::from_str(&manifest) {
+
+    let manifest = match toml::from_str(&manifest) {
         Ok(o) => Ok(o),
         Err(e) => Err(format!("Error parsing manifest: {}.", e)),
+    }?;
+
+    validate_manifest(manifest)
+}
+
+// Using (https://github.com/rust-lang/cargo/blob/master/src/cargo/util/toml/mod.rs#L570 +
+// https://github.com/rust-lang/cargo/blob/master/src/cargo/ops/cargo_new.rs#L166) for reference
+
+fn validate_name(name: &str, use_case: &str) -> Result<(), String> {
+    // if true returns formatted error
+    restricted_names::contains_invalid_char(name, use_case)?;
+
+    if restricted_names::is_keyword(name) {
+        return Err(format!(
+            "the name `{name}` cannot be used as a package name, it is a Sway keyword"
+        ));
+    }
+    if restricted_names::is_conflicting_artifact_name(name) {
+        return Err(format!(
+            "the name `{name}` cannot be used as a package name, \
+            it conflicts with forc's build directory names"
+        ));
+    }
+    if name == "test" {
+        return Err("the name `test` cannot be used as a package name, \
+            it conflicts with Sway's built-in test library"
+            .into());
+    }
+    if restricted_names::is_conflicting_suffix(name) {
+        return Err(format!(
+            "the name `{name}` is part of Sway's standard library\n\
+            It is recommended to use a different name to avoid problems."
+        ));
     }
+    if restricted_names::is_windows_reserved(name) {
+        if cfg!(windows) {
+            return Err(format!(
+                "cannot use name `{name}`, it is a reserved Windows filename"
+            ));
+        } else {
+            return Err(format!(
+                "the name `{name}` is a reserved Windows filename\n\
+                This package will not work on Windows platforms."
+            ));
+        }
+    }
+    if restricted_names::is_non_ascii_name(name) {
+        return Err(format!(
+            "the name `{name}` contains non-ASCII characters\n\
+            Support for non-ASCII crate names is experimental and only valid \
+            on Rust's nightly toolchain."
```

Comment:
Rust toolchain? 😂

If we're not going to support, say forbidden.

Reply:
gotcha! Most of this was from Cargo, that I repurposed. Thanks!